### ISSUE TITLE
pylint fails to detect missing method in concrete class implementing an abstract base class

### ISSUE THREAD

**Issue by dstromberg** [PyCQA/pylint#1673]
### Steps to reproduce
1. svn checkout http://stromberg.dnsalias.org/svn/pylint-interface-experiment/trunk/
2. Read README
3. make
4. Study SSCCE obtained in step 1

### Current behavior
Pylint passes the code without a warning, but running the code gives TypeError: Can't instantiate abstract class PrintThing with abstract methods print_thing

### Expected behavior
Pylint should emit a warning indicating that the print_thing() method is missing from py3prn.PrintThing.

### pylint --version output
$ /usr/bin/pylint3 --version
below cmd output started 2017 Sat Sep 23 12:49:14 PM PDT
pylint3 1.5.2,
astroid 1.4.4
Python 3.5.2 (default, Nov 17 2016, 17:05:23)
[GCC 5.4.0 20160609]

It also happens with:
$ /usr/local/cpython-3.4/bin/pylint --version
below cmd output started 2017 Sat Sep 23 12:49:57 PM PDT
pylint 1.7.1,
astroid 1.5.2
Python 3.4.2 (default, Apr 17 2017, 09:05:12)
[GCC 5.4.0 20160609]

Thanks for pylint! I love it.

**Comment by dstromberg** [comment]
BTW, if there's some other way of having pylint check two implementations of a single interface, that'd be good too.

**Comment by PCManticore** [comment]
You can also define a class using a metaclass with ``six``, which should make your code easier to grasp by *pylint*:

```
import six

@six.add_metaclass(abc.ABCMeta)
class SomeClass(object):
   ...
```

In your examples, there are two problems which inhibit pylint's capability of understanding your program:

1. first is the metaclass defined in an ``if`` statement. Currently pylint has troubles inferring objects encapsulated in flow control statements, which is a thing that we'll improve in the future.

2. where pylint is getting tripped is in the conditional definition of your imports. In this case, pylint will try to infer the class object PrintThing from both definitions of ``prn``, since it lacks inferring on which branch it should proceed. Since py2prn has a PrintThing that is not abstract, pylint will happily infer that PrintThing from py3prn is not abstract as well, leading to this issue.

**Comment by PCManticore** [comment]
Alright, so the culprit was that we were using the first inferred object for a class, e.g ``next(node.infer())``. Changed to using all inferred objects and this error is now caught (albeit without the problem with defining conditionally `abc.ABC`). That should probably be handled using the `six` snippet I gave you, which will make pylint happy.

**Comment by dstromberg** [comment]
That's great news.

I modified http://stromberg.dnsalias.org/svn/pylint-interface-experiment/trunk/prnabc.py to use six.

Is there a prerelease pylint somewhere I can try, to help test this improved inference?

Thanks!

**Comment by rogalski** [comment]
@dstromberg [How to state in requirements.txt a direct github source](https://stackoverflow.com/questions/16584552/how-to-state-in-requirements-txt-a-direct-github-source) should work just fine.

**Comment by rogalski** [comment]
> Is there a way of doing that within the same virtualenv?

Virtualenv is equivalent to one-and-only-one Python version associated with it, you don't need multiple entry points in that case. For multiple interpreters, I usually go with something like `python -m pylint` and `python3 -m pylint`. About `pylint3` entry point, there were a ticket some time ago, but it was rejected - #1366
